src/Common/src/System/Net/Http/TlsCertificateExtensions.cs

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+
+namespace System.Net.Http
+{
+    /// <summary>
+    /// Defines Extension methods which are meant to be re-used across WinHttp & UnixHttp Handlers
+    /// </summary>
+    internal static class TLSCertificateExtensions
+    {
+        private const string ClientCertificateOid = "1.3.6.1.5.5.7.3.2";
+
+        /// <summary>
+        ///   returns true if the X509 Certificate can be used  as SSL Client Certificate.
+        /// </summary>
+        private static bool IsClientCertificate(X509Certificate2 cert)
+        {
+            Debug.Assert(cert != null, "certificate cannot be null");
+
+            bool hasClientCertExt = false;
+            bool forDigitalSignature = false;
+            bool noExtendedRestrictions = true;
+            bool noUsageRestrictions = true;
+            foreach(X509Extension extension in cert.Extensions)
+            {
+                // check if the extension is an enhanced usage ext.
+                // But do this only if needed. No point going over it, if we already have established that our cert has the
+                // required extension.
+                if (!hasClientCertExt)
+                {
+                    X509EnhancedKeyUsageExtension enhancedUsageExt = extension as X509EnhancedKeyUsageExtension;
+                    if (enhancedUsageExt != null)
+                    {
+                        noExtendedRestrictions = false;
+                        foreach(Oid oid in enhancedUsageExt.EnhancedKeyUsages)
+                        {
+                            if (string.Equals(ClientCertificateOid, oid.Value))
+                            {
+                                hasClientCertExt = true;
+                                continue;
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    Debug.Assert(hasClientCertExt && !noExtendedRestrictions, "client Cert extension if present, is only as an EKU");
```

Comment:
I also don't understand the assert (and/or the message).

A certificate can have hundreds of extensions.  The EKU won't be the only one.  And if you mean "it should have been specially decoded as the X509EnhancedKeyUsageExtension type", well... that's how you found it.

So I'm really not sure what this is saying.